### ISSUE TITLE
Miscompilation with NaN values and "-O3 -ffinite-math-only"

### ISSUE THREAD

**Issue by AMS21** [emscripten-core/emscripten#15743]
The following source code should always print `NaN` without `-ffinite-math-only` and `Not NaN` with. But emcc always prints `NaN`.
```cpp
#include <cstdio>

int main() {
  volatile float n = 0.0 / 0.0;
  if (n != n) {
    std::printf("NaN\n");
    return 1;
  }

  std::printf("Not NaN\n");
  return 0;
}
```
command line options: `emcc -O3 source.cpp -o emcc.js`
[godbolt link](https://godbolt.org/z/YEno8dKhv)

`emcc -v`
> emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 2.0.27 (7e538a419c3649f3a540a57beab347aa8f6c6271)
> clang version 14.0.0 (/srcdest/llvm-project 78e87970af888bbbd5652c31f3a8454e8e9dd5b8)
> Target: wasm32-unknown-emscripten
> Thread model: posix
> InstalledDir: /opt/emscripten-llvm/bin

**Comment by sbc100** [comment]
Are you sure the its not reasonable for `-ffinite-math-only` to *not* do this optimization?

From the man page for gcc:

```
       -ffinite-math-only
           Allow optimizations for floating-point arithmetic that assume that arguments
           and results are not NaNs or +-Infs.
```

From my reading it sounds like doing nothing at all is reasonable implementation of this flag.

There may well be a missed optimization opportunity though.. is that what you mean?

**Comment by AMS21** [comment]
Yes, when we assume there are non NaNs `x != x` should always be false just like with inter values.

**Comment by AMS21** [comment]
Corrected the title and wording. Was a bit late yesterday and I wanted to simply get the issue report done.
Sorry for the inconvenience
